### PR TITLE
finalise deposits before we initialise the node

### LINKED ISSUE #9610: Missing proposal because of  DepositsNumberAndRootAtHeight
# 💎 Issue

### Background

As part of debug session for the GetBlock() optimisation, this issue was found and needs to be looked in to.

### Description

**GetBlock() -> eth1DataMajorityVote() -> DepositsNumberAndRootAtHeight()**

- **DepositsNumberAndRootAtHeight()** - This clearly the **biggest bang for the buck**. Even though the trace shows up here in the trace, the log shows that the OnBlock()->insertFinalizedDeposits() is the actual culprit.
- **scenario 1** 

- **scenario 2**

- **scenario 3**

### PR DESCRIPTION
**Pull request by jmozah**
**What type of PR is this?**
 Bug fix

**What does this PR do? Why is it needed?**

The PR  finalizes the accumulated deposits before initiazing the node.
when a node pauses for some time and starts again, the deposits to finalize
accumulates. we finalize them here before we are ready to receive a block.
Otherwise, the first few block will not do their duties properly as we will
hold the lock and be busy finalizing the deposits.

**Which issues(s) does this PR fix?**

Fixes #9610 

**Other notes for review**

### CODE REVIEW

File: beacon-chain/blockchain/service.go [hunk 1]
```diff
@@ -194,6 +194,15 @@ func (s *Service) Start() {
 			log.Fatalf("Could not verify weak subjectivity checkpoint: %v", err)
 		}
 
+		// when a node pauses for some time and starts again, the deposits to finalize
+		// accumulates. we finalize them here before we are ready to receive a block.
+		// Otherwise, the first few block will not do their duties properly as we will
+		// hold the lock and be busy finalizing the deposits.
+		fRoot := bytesutil.ToBytes32(s.finalizedCheckpt.Root)
+		if err := s.insertFinalizedDeposits(s.ctx, fRoot); err != nil {
+			log.WithError(err).Error("Could not insert finalized deposits.")
```
Comment (terencechain) [inline beacon-chain/blockchain/service.go:203]:
```suggestion
			log.WithError(err).Error("Could not insert finalized deposits")
```

File: beacon-chain/blockchain/service.go [hunk 2]
```diff
@@ -194,6 +194,15 @@ func (s *Service) Start() {
 			log.Fatalf("Could not verify weak subjectivity checkpoint: %v", err)
 		}
 
+		// when a node pauses for some time and starts again, the deposits to finalize
+		// accumulates. we finalize them here before we are ready to receive a block.
+		// Otherwise, the first few block will not do their duties properly as we will
```
Comment (terencechain) [inline beacon-chain/blockchain/service.go:199]:
```suggestion
		// Otherwise, the first few blocks will be slower to compute as we will
```

**Review by rauljordan — APPROVED** [review]

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
Please add some kind of regression test for this new logic.

**Review by nisdas — COMMENTED** [review]
Hey looking at this a bit more, if we insert the finalized deposits here, we will end up doing 3 state fetches during startup. The fetches are expensive because we have to recompute the state from an archived point. It might be more efficient to simply insert it here in `powchain`.
https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/powchain/service.go#L628

We already fetch the finalized state there so we can re-use that when inserting finalized deposits. Also makes this easier to test.

**Comment by jmozah** [comment]
> Hey looking at this a bit more, if we insert the finalized deposits here, we will end up doing 3 state fetches during startup. The fetches are expensive because we have to recompute the state from an archived point. It might be more efficient to simply insert it here in `powchain`.
> https://github.com/prysmaticlabs/prysm/blob/develop/beacon-chain/powchain/service.go#L618
> 
> We already fetch the finalized state there so we can re-use that when inserting finalized deposits. Also makes this easier to test.

@nisdas I considered doing it there first as we load the caches there.
There are two problems with doing it there (infact loading caches too has a problem now).

1) The system should not send `state.Initialized` before we have 
     - loaded all the cache
     - Finalised all of the caches. (this is the current problem though)
    This can be controlled only in blockchain.service.
    
 2) We don't have access to

**Review by nisdas — APPROVED** [review]

**Comment by nisdas** [comment]
@prestonvanloon , ping on another review for the PR.

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**